docs/installation/configuration.rst
=============
Configuration
=============

Create and configure a new Django project
=========================================

If you don't have a django project yet create one now:

.. code-block:: bash
    
    $ django-admin.py startproject <YOUR_PROJECT_NAME>
    $ cd <YOUR_PROJECT_NAME>

Edit project's settings.py
==========================

1. Configure database (SQLite example):
    
  .. code-block:: python
        
        DATABASES = {
            'default': {
                'ENGINE': 'django.db.backends.sqlite3',   # Add 'postgresql_psycopg2', 'postgresql', 'mysql', 'sqlite3' or 'oracle'.
                'NAME': '/Users/dkerwin/dev/dpki/pki.db', # Or path to database file if using sqlite3.
                'USER': '',                      # Not used with sqlite3.
                'PASSWORD': '',                  # Not used with sqlite3.
                'HOST': '',                      # Set to empty string for localhost. Not used with sqlite3.
                'PORT': '',                      # Set to empty string for default. Not used with sqlite3.
            }
        }

2. Add django-pki template directory to TEMPLATE_DIRS if 'django.template.loaders.app_directories.Loader' is not in TEMPLATE_LOADERS:
    
  .. code-block:: python
        
        TEMPLATE_LOADERS = (
            'django.template.loaders.filesystem.Loader',
            'django.template.loaders.app_directories.Loader',
        )

  **or**
  
  .. code-block:: python
        
        TEMPLATE_DIRS = ('/Library/Python/2.6/site-packages/pki/templates',)


3. Add 'pki.middleware.PkiExceptionMiddleware' to MIDDLEWARE_CLASSES (used for exception logging):
    
  .. code-block:: python
        
        MIDDLEWARE_CLASSES = (
            '...',
            'pki.middleware.PkiExceptionMiddleware',
        )

4. Add 'django.contrib.admin', 'south' and 'pki' to INSTALLED_APPS:

  .. warning:: Make sure **pki** is specified after **south** as unit tests won't work otherwise

  .. code-block:: python
        
        INSTALLED_APPS = (
            '...',
            'django.contrib.admin',
            'south',
            'pki',
        )

5. Set MEDIA_URL and MEDIA_ROOT
    
  The values of MEDIA_URL and MEDIA_ROOT depend on your configuration.
  MEDIA_ROOT is the filesystem path to the django-pki media files (<PATH_TO_DJANGO_PKI>/media). You can of cause copy or symlink the files to another location.
  MEDIA_URL is the URL part where the media files can be accessed. Here are some examples:

  .. code-block:: python

    MEDIA_ROOT = '/Library/Python/2.6/site-packages/pki/media/'
    MEDIA_ROOT = '/var/www/myhost/static/pki'
    
    MEDIA_URL = '/static/'
    MEDIA_URL = '/pki_media/'

6. Set ADMIN_MEDIA_PREFIX

Configure django-pki settings (in projects settings.py)
=======================================================

You can use any combination of the following parameters:

**PKI_DIR** (*Default = /path-to-django-pki/PKI; Type = Python String*)
    Absolute path to directory for pki storage. Must be writable

**PKI_OPENSSL_BIN** (*Default = /usr/bin/openssl; Type = Python String*)
    Path to openssl binary

**PKI_OPENSSL_CONF** (*Default = PKI_DIR/openssl.conf; Type = Python String*)
    Location of OpenSSL config file (openssl.conf)

**PKI_OPENSSL_TEMPLATE** (*Default = pki/openssl.conf.in; Type = Python String*)
    OpenSSL configuration template (Shouldn't be changen unless really neccessary)

**PKI_LOG** (*Default = PKI_DIR/pki.log; Type = Python String*)
    Full qualified path to logfile for PKI actions

**PKI_LOGLEVEL** (*Default = info; Type = Python String*)
    Logging level according to Python logging module

**JQUERY_URL** (*Default = pki/jquery-1.4.2.min.js; Type = Python String*)
    Alternative jQuery url

**PKI_SELF_SIGNED_SERIAL** (*Default = 0x0; Type = Python Number*)
    The serial of self-signed certificates. Set to 0 or 0x0 to get a random number (0xabc = HEX; 123 = DEC)

**PKI_DEFAULT_KEY_LENGTH** (*Default = 1024; Type = Python Number*)
    The default key length

**PKI_DEFAULT_COUNTRY** (*Default = DE; Type = Python String*)
    The default country (as 2-letter code) selected (http://en.wikipedia.org/wiki/ISO_3166-1_alpha-2)

**PKI_PASSPHRASE_MIN_LENGTH** (*Default = 8; Type = Python Number*)
    The minimum length for passphrases

**PKI_ENABLE_GRAPHVIZ** (*Default = False; Type = Python Boolean*)
    Enable graphviz support (see requirements)

**PKI_GRAPHVIZ_DIRECTION** (*Default = LR; Type = Python String*)
    Graph tree direction (LR=left-to-right, TD=top-down)

**PKI_ENABLE_EMAIL** (*Default = False; Type = Python Boolean*)
    Email delivery to certificate's email address. May require additional `Django paramters (EMAIL_*) <http://docs.djangoproject.com/en/dev/ref/settings/>`_

**Example:**
::
    
    ## django-pki specific parameters
    PKI_DIR = '/var/pki/ssl_store'
    PKI_OPENSSL_BIN = '/opt/openssl/bin/openssl'
    PKI_OPENSSL_CONF = '/opt/openssl/bin/etc/openssl.conf'
    PKI_LOG = '/var/log/django-pki.log'
    PKI_LOGLEVEL = 'error'
    JQUERY_URL = 'http://static.company.com/js/jquery.js'
    PKI_SELF_SIGNED_SERIAL = 0x0
    PKI_DEFAULT_KEY_LENGTH = 2048
    PKI_DEFAULT_COUNTRY = 'UK'
    PKI_PASSPHRASE_MIN_LENGTH = 12
    PKI_ENABLE_GRAPHVIZ = True
    PKI_GRAPHVIZ_DIRECTION = 'TD'
    PKI_ENABLE_EMAIL = True
    
    ## django specific email configuration
    EMAIL_HOST = "192.168.1.1"
    EMAIL_HOST_USER = "relayuser"
    EMAIL_HOST_PASSWORD = "icanrelay"
    DEFAULT_FROM_EMAIL = "pki@my-company.com"

Configure projects urls.py
==========================

1. Enable admin application::
    
    from django.contrib import admin 
    admin.autodiscover()

2. Add exception handler::
    
    handler500 = 'pki.views.show_exception'

3. Add the following lines to urlpatterns

  ::
    
    (r'^admin/', include(admin.site.urls)),
    (r'^', include('pki.urls', 'pki')),

4. If you want to serve static files with ``./manage.py runserver`` in DEBUG mode, add the following code:
    
  .. warning:: **!! Do not use this in production !!**
    
  ::

    from django.conf import settings
    
    if settings.DEBUG:
        M = settings.MEDIA_URL
        if M.startswith('/'): M = M[1:]
        if not M.endswith('/'): M += '/'
        urlpatterns += patterns('', (r'^%s(?P<path>.*)$' % M, 'django.views.static.serve',
                                {'document_root': settings.MEDIA_ROOT}))

Initialize database
===================

* Initialize database::
    
    $ python manage.py syncdb
    Syncing...
    Creating table auth_permission
    Creating table auth_group_permissions
    Creating table auth_group
    Creating table auth_user_user_permissions
    Creating table auth_user_groups
    Creating table auth_user
    Creating table auth_message
    Creating table django_content_type
    Creating table django_session
    Creating table django_site
    Creating table django_admin_log
    Creating table south_migrationhistory
    
    You just installed Django's auth system, which means you don't have any superusers defined.
    Would you like to create one now? (yes/no): yes
    Username (Leave blank to use 'dkerwin'): admin
    E-mail address: a@b.com
    Password: 
    Password (again): 
    Superuser created successfully.
    Installing index for auth.Permission model
    Installing index for auth.Group_permissions model
    Installing index for auth.User_user_permissions model
    Installing index for auth.User_groups model
    Installing index for auth.Message model
    Installing index for admin.LogEntry model
    No fixtures found.
    
    Synced:
     > django.contrib.auth
     > django.contrib.contenttypes
     > django.contrib.sessions
     > django.contrib.sites
     > django.contrib.messages
     > django.contrib.admin
     > debug_toolbar
     > south
    
    Not synced (use migrations):
     - pki
    (use ./manage.py migrate to migrate these)

* Create django-pki tables. This is a south migration::
    
    $ python manage.py migrate pki
    Running migrations for pki:
     - Migrating forwards to 0003_auto__add_pkichangelog.
     > pki:0001_initial
     > pki:0002_auto__add_field_certificateauthority_crl_distribution
     > pki:0003_auto__add_pkichangelog
     - Loading initial data for pki.
    No fixtures found.


PKI store layout (PKI_DIR)
==========================

Every certificate authority (CA) lives in a dedicated directory in PKI_DIR. There is a special directory named "_SELF_SIGNED_CERTIFICATES" which
contains all self-signed non-CA certificates. A certificate authority directory contains the follwoing sub-directories and files:

* Directories:
    * private: Private key of the CA
    * crl: Generated CRL
    * certs: All direct related certificates (subCA certificates or end-user certificates when it's a edge CA).
      Symlinks between the serialnumber and the hash are created for every certificate.
* Files:
    * index.txt(.old): The CA index
    * index.txt.attr(.old): Extra attribtes for index.txt
    * serial(.old): Current CA serial number
    * crlnumber(.old): Current CRL serial number
    * [CA NAME]-chain.cert.pem: The CA chain including the own CA certificate

Example::
    
    Root_CA/
    Root_CA/certs
    Root_CA/certs/01.pem
    Root_CA/certs/02.pem
    Root_CA/certs/518c671c.0
    Root_CA/certs/771a33d0.0
    Root_CA/certs/Root_CA.cert.pem
    Root_CA/crl
    Root_CA/crl/Root_CA.crl.pem
    Root_CA/crlnumber
    Root_CA/crlnumber.old
    Root_CA/index.txt
    Root_CA/index.txt.attr
    Root_CA/index.txt.attr.old
    Root_CA/index.txt.old
    Root_CA/private
    Root_CA/private/Root_CA.key.pem
    Root_CA/Root_CA-chain.cert.pem
    Root_CA/serial
    Root_CA/serial.old
    _SELF_SIGNED_CERTIFICATES/
    _SELF_SIGNED_CERTIFICATES/certs
    _SELF_SIGNED_CERTIFICATES/certs/selfsigned1.cert.pem
    _SELF_SIGNED_CERTIFICATES/certs/selfsigned1.key.pem



_sources/installation/configuration.txt
=============
Configuration
=============

Create and configure a new Django project
=========================================

If you don't have a django project yet create one now:

.. code-block:: bash
    
    $ django-admin.py startproject <YOUR_PROJECT_NAME>
    $ cd <YOUR_PROJECT_NAME>

Edit project's settings.py
==========================

1. Configure database (SQLite example):
    
  .. code-block:: python
        
        DATABASES = {
            'default': {
                'ENGINE': 'django.db.backends.sqlite3',   # Add 'postgresql_psycopg2', 'postgresql', 'mysql', 'sqlite3' or 'oracle'.
                'NAME': '/Users/dkerwin/dev/dpki/pki.db', # Or path to database file if using sqlite3.
                'USER': '',                      # Not used with sqlite3.
                'PASSWORD': '',                  # Not used with sqlite3.
                'HOST': '',                      # Set to empty string for localhost. Not used with sqlite3.
                'PORT': '',                      # Set to empty string for default. Not used with sqlite3.
            }
        }

2. Add django-pki template directory to TEMPLATE_DIRS if 'django.template.loaders.app_directories.Loader' is not in TEMPLATE_LOADERS:
    
  .. code-block:: python
        
        TEMPLATE_LOADERS = (
            'django.template.loaders.filesystem.Loader',
            'django.template.loaders.app_directories.Loader',
        )

  **or**
  
  .. code-block:: python
        
        TEMPLATE_DIRS = ('/Library/Python/2.6/site-packages/pki/templates',)


3. Add 'pki.middleware.PkiExceptionMiddleware' to MIDDLEWARE_CLASSES (used for exception logging):
    
  .. code-block:: python
        
        MIDDLEWARE_CLASSES = (
            '...',
            'pki.middleware.PkiExceptionMiddleware',
        )

4. Add 'django.contrib.admin', 'south' and 'pki' to INSTALLED_APPS:

  .. warning:: Make sure **pki** is specified after **south** as unit tests won't work otherwise

  .. code-block:: python
        
        INSTALLED_APPS = (
            '...',
            'django.contrib.admin',
            'south',
            'pki',
        )

5. Set MEDIA_URL and MEDIA_ROOT
    
  The values of MEDIA_URL and MEDIA_ROOT depend on your configuration.
  MEDIA_ROOT is the filesystem path to the django-pki media files (<PATH_TO_DJANGO_PKI>/media). You can of cause copy or symlink the files to another location.
  MEDIA_URL is the URL part where the media files can be accessed. Here are some examples:

  .. code-block:: python

    MEDIA_ROOT = '/Library/Python/2.6/site-packages/pki/media/'
    MEDIA_ROOT = '/var/www/myhost/static/pki'
    
    MEDIA_URL = '/static/'
    MEDIA_URL = '/pki_media/'

6. Set ADMIN_MEDIA_PREFIX

Configure django-pki settings (in projects settings.py)
=======================================================

You can use any combination of the following parameters:

**PKI_DIR** (*Default = /path-to-django-pki/PKI; Type = Python String*)
    Absolute path to directory for pki storage. Must be writable

**PKI_OPENSSL_BIN** (*Default = /usr/bin/openssl; Type = Python String*)
    Path to openssl binary

**PKI_OPENSSL_CONF** (*Default = PKI_DIR/openssl.conf; Type = Python String*)
    Location of OpenSSL config file (openssl.conf)

**PKI_OPENSSL_TEMPLATE** (*Default = pki/openssl.conf.in; Type = Python String*)
    OpenSSL configuration template (Shouldn't be changen unless really neccessary)

**PKI_LOG** (*Default = PKI_DIR/pki.log; Type = Python String*)
    Full qualified path to logfile for PKI actions

**PKI_LOGLEVEL** (*Default = info; Type = Python String*)
    Logging level according to Python logging module

**JQUERY_URL** (*Default = pki/jquery-1.4.2.min.js; Type = Python String*)
    Alternative jQuery url

**PKI_SELF_SIGNED_SERIAL** (*Default = 0x0; Type = Python Number*)
    The serial of self-signed certificates. Set to 0 or 0x0 to get a random number (0xabc = HEX; 123 = DEC)

**PKI_DEFAULT_COUNTRY** (*Default = DE; Type = Python String*)
    The default country (as 2-letter code) selected (http://en.wikipedia.org/wiki/ISO_3166-1_alpha-2)

**PKI_PASSPHRASE_MIN_LENGTH** (*Default = 8; Type = Python Number*)
    The minimum length for passphrases

**PKI_ENABLE_GRAPHVIZ** (*Default = False; Type = Python Boolean*)
    Enable graphviz support (see requirements)

**PKI_GRAPHVIZ_DIRECTION** (*Default = LR; Type = Python String*)
    Graph tree direction (LR=left-to-right, TD=top-down)

**PKI_ENABLE_EMAIL** (*Default = False; Type = Python Boolean*)
    Email delivery to certificate's email address. May require additional `Django paramters (EMAIL_*) <http://docs.djangoproject.com/en/dev/ref/settings/>`_

**Example:**
::
    
    ## django-pki specific parameters
    PKI_DIR = '/var/pki/ssl_store'
    PKI_OPENSSL_BIN = '/opt/openssl/bin/openssl'
    PKI_OPENSSL_CONF = '/opt/openssl/bin/etc/openssl.conf'
    PKI_LOG = '/var/log/django-pki.log'
    PKI_LOGLEVEL = 'error'
    JQUERY_URL = 'http://static.company.com/js/jquery.js'
    PKI_SELF_SIGNED_SERIAL = 0x0
    PKI_DEFAULT_COUNTRY = 'UK'
    PKI_PASSPHRASE_MIN_LENGTH = 12
    PKI_ENABLE_GRAPHVIZ = True
    PKI_GRAPHVIZ_DIRECTION = 'TD'
    PKI_ENABLE_EMAIL = True
    
    ## django specific email configuration
    EMAIL_HOST = "192.168.1.1"
    EMAIL_HOST_USER = "relayuser"
    EMAIL_HOST_PASSWORD = "icanrelay"
    DEFAULT_FROM_EMAIL = "pki@my-company.com"

Configure projects urls.py
==========================

1. Enable admin application::
    
    from django.contrib import admin 
    admin.autodiscover()

2. Add exception handler::
    
    handler500 = 'pki.views.show_exception'

3. Add the following lines to urlpatterns

  ::
    
    (r'^admin/', include(admin.site.urls)),
    (r'^', include('pki.urls', 'pki')),

4. If you want to serve static files with ``./manage.py runserver`` in DEBUG mode, add the following code:
    
  .. warning:: **!! Do not use this in production !!**
    
  ::

    from django.conf import settings
    
    if settings.DEBUG:
        M = settings.MEDIA_URL
        if M.startswith('/'): M = M[1:]
        if not M.endswith('/'): M += '/'
        urlpatterns += patterns('', (r'^%s(?P<path>.*)$' % M, 'django.views.static.serve',
                                {'document_root': settings.MEDIA_ROOT}))

Initialize database
===================

* Initialize database::
    
    $ python manage.py syncdb
    Syncing...
    Creating table auth_permission
    Creating table auth_group_permissions
    Creating table auth_group
    Creating table auth_user_user_permissions
    Creating table auth_user_groups
    Creating table auth_user
    Creating table auth_message
    Creating table django_content_type
    Creating table django_session
    Creating table django_site
    Creating table django_admin_log
    Creating table south_migrationhistory
    
    You just installed Django's auth system, which means you don't have any superusers defined.
    Would you like to create one now? (yes/no): yes
    Username (Leave blank to use 'dkerwin'): admin
    E-mail address: a@b.com
    Password: 
    Password (again): 
    Superuser created successfully.
    Installing index for auth.Permission model
    Installing index for auth.Group_permissions model
    Installing index for auth.User_user_permissions model
    Installing index for auth.User_groups model
    Installing index for auth.Message model
    Installing index for admin.LogEntry model
    No fixtures found.
    
    Synced:
     > django.contrib.auth
     > django.contrib.contenttypes
     > django.contrib.sessions
     > django.contrib.sites
     > django.contrib.messages
     > django.contrib.admin
     > debug_toolbar
     > south
    
    Not synced (use migrations):
     - pki
    (use ./manage.py migrate to migrate these)

* Create django-pki tables. This is a south migration::
    
    $ python manage.py migrate pki
    Running migrations for pki:
     - Migrating forwards to 0003_auto__add_pkichangelog.
     > pki:0001_initial
     > pki:0002_auto__add_field_certificateauthority_crl_distribution
     > pki:0003_auto__add_pkichangelog
     - Loading initial data for pki.
    No fixtures found.


PKI store layout (PKI_DIR)
==========================

Every certificate authority (CA) lives in a dedicated directory in PKI_DIR. There is a special directory named "_SELF_SIGNED_CERTIFICATES" which
contains all self-signed non-CA certificates. A certificate authority directory contains the follwoing sub-directories and files:

* Directories:
    * private: Private key of the CA
    * crl: Generated CRL
    * certs: All direct related certificates (subCA certificates or end-user certificates when it's a edge CA).
      Symlinks between the serialnumber and the hash are created for every certificate.
* Files:
    * index.txt(.old): The CA index
    * index.txt.attr(.old): Extra attribtes for index.txt
    * serial(.old): Current CA serial number
    * crlnumber(.old): Current CRL serial number
    * [CA NAME]-chain.cert.pem: The CA chain including the own CA certificate

Example::
    
    Root_CA/
    Root_CA/certs
    Root_CA/certs/01.pem
    Root_CA/certs/02.pem
    Root_CA/certs/518c671c.0
    Root_CA/certs/771a33d0.0
    Root_CA/certs/Root_CA.cert.pem
    Root_CA/crl
    Root_CA/crl/Root_CA.crl.pem
    Root_CA/crlnumber
    Root_CA/crlnumber.old
    Root_CA/index.txt
    Root_CA/index.txt.attr
    Root_CA/index.txt.attr.old
    Root_CA/index.txt.old
    Root_CA/private
    Root_CA/private/Root_CA.key.pem
    Root_CA/Root_CA-chain.cert.pem
    Root_CA/serial
    Root_CA/serial.old
    _SELF_SIGNED_CERTIFICATES/
    _SELF_SIGNED_CERTIFICATES/certs
    _SELF_SIGNED_CERTIFICATES/certs/selfsigned1.cert.pem
    _SELF_SIGNED_CERTIFICATES/certs/selfsigned1.key.pem



CHANGELOG
0.20.0
======
a6e2294 - Failed confilct solving
805f662 - Merge branch 'next'
abc07e7 - Fix download zipfile content
e078c4d - Fix pkcs12 export when key has a passphrase set;
56c2c7b - Blacklist support to deny ca names; JS fix for x509 form submit;
8759d41 - get_pki_icon_html moved to helper.py; Centered CRL distribution point in changelist; Fix error in self-signed cert creation; extension filter; New unit test
1509134 - Fix orm error in migration 0006; New migration for crl distribution points; clr_dpoints in base model; Form validation for crl_dpoints; Fix CRL DP validation; Raw cert parsing unittest
cbcfe6f - New setting PKI_DEFAULT_KEY_LENGTH; Add string import; More unittests; Fix permission handling in download view
feed12a - Add missing 0.10.5 changelog; New setting PKI_PASSPHRASE_MIN_LENGTH; Requirement for django>=1.2; Add docs to dist package; Use of model validators to simplify form validation; Remove unused imports; View redirects using urlresolver;
73144db - Setting version to 0.20.0; Unittests; Updated docs
ca27b4c - More unittests; Removed not working C_D_P code;
4240842 - Prevent x509Extension model save; Override PKI_OPENSSL_CONF; Switch to TestCase
027a0fb - Finally some unittests; Fix revoke handling in CA mode; Removal of all affected CA folders in delete mode; Raise exception if action is invalid in certificate save; Make ca_dir of Openssl available;
f598c94 - Fix migration 0011; ContentType needs to be populated manually; test command uses south test command; paginate 25 in admin.py;
539fef6 - Documentation of upgrade procedure; Migration 0006 needs orm to get information of old fields;
6b83670 - Documentation on PKI_DIR layout; Fix js match for model; Useage of getattr for user in model save to handle missing users; Fix encoding and description changes in update;
3f8cc46 - Link to project page
d6b861d - Increased readability; Documentation about keyUsage and extendedKeyUsage; Migration to add download permission for PKI objects; Unique key for keyusage and extendedkeyusage; Set user to UNKNOWN if not given; Add error when revoking a self-signed enduser certificate;
3c31dbe - Replace templatetag model_for_object by model_for_content_type as this works on all models; Seperate js blocks for all models;
360aac8 - Double underscore in changlog template
983538d - Switch to messges framework. Replacement for message_set
76ee0d4 - Set x509 extensions to disabled via js to prevent modifications
b10bc96 - Renamed object_history.html to object_changelogs.html; delete_confirmation.html uses default block when object is not ca or cert; Removed special admin urls as handled in admin.py; x509 extension in delete details; Fix exception when x509 form validation when no name was given; Prevent modifications of x509 table from the admin;
c150ae0 - Remove db fields "type" and "pem_encoded"; Fix extension handling in certificate signing;
5422bfa - Manage x509 extensions from the admin. This includes 3 new tables, fixtures to make the old certificates still valid, changes foreignkey queries, new openssl.conf template, ...; Removed db fields: pf_encrypted, subcas_allowed, cert_extension; Updated navigation;
0291bb5 - Updated docs; Updated south migration; Column cycle for history view; Combined deletion form; Simplified image display; Model functions for changelog management
b98fb17 - Replace builtin admin history. New model PkiChangelog. Override admin save_model funtions to pass username to model save. Partial changelog integration in model.save. Change jquery version in delete_confirmation.
8161ca5 - Show certificate state in change form
230289c - Set revoke status in DB for all child objects
35da21e - Renamed active_center to Valid_center; Changed Expiry_date from div to span and use line-through" when item is revoked; Fix graphviz error when parent is revoked;
32c6c66 - Added clock image and add to CA clock. Changed action descriptions. Remove nokeys from pkcs12 generation
966eed6 - Switch to jquery 1.5; Show number of CAs when metadata is refreshed; Add library kinks to docs
e580497 - Special extension for self-signed certificates; Refresh metadata navigation link and view;
7a13503 - Tipsy for certificates too; Fix graphviz tree for RootCa's; get_icon_html removed and merged into get_pki_icon_html;
dedb904 - Rename OpensslActions -> Openssl; Tipsy js integration; More logging in openssl.py
1a10f61 - Updated documentation
7923568 - Switched theme. Deletion and exclusion of docs/_build
2fba0ec - Doc update
17f7aac - Documentation source and changed CRL DP help text
44e969b - Merge branch 'gh-pages' of github.com:dkerwin/django-pki into next
ae62533 - First pages commit
92b4ad9 - Layout changes; Fix form validation in self-signed mode; Switch to south for db migration; New DB column and  x509 extensions for CRL distribution points; Fix delete view bug; Updated README
d3f34ae - CA clock to display the current time on the PKI server. Fetched during template rendering and updated without server communication. New templatetag to set initial time
e07eab1 - New templatetag for MEDIA_URL. Form validation fails when name is used in the PKI store
97beaa1 - Switched to named URLs
4e65a5b - Harmonized urls (type -> model). Redirect to admin index from 500.html
2f3d2ea - Fix image display
3bc7cf0 - Icons and changelist link to filter associated certificate in a new window. Order field for Creation_date. Replaced a lot of useless <center> tags
0d935c5 - Updated changelog
df9aad6 - More css. Faked test command to automatically set the right setings file
ad0c7b4 - Fix BASE_URL problems
d7ea863 - Minified js. Getting BASE_URL from get_script_prefix().
daf0d7e - Fixed settings.py js path
d12b5f9 - Relocated css and js files.
4914c03 - Custom css file. Templates for all app related views. New template tag to create navigation menu. Fix delete problem with initial_id variable
7ebc8f4 - Skip ca-chain for self-signed non CA certificates when zip is created

pki/settings.py
##############################################################
## PKI settings
##############################################################

import os
from django.conf import settings
from django.core.urlresolvers import get_script_prefix

PKI_APP_DIR = os.path.abspath(os.path.dirname(__file__))

# blacklisted CA names
PKI_CA_NAME_BLACKLIST = ('_SELF_SIGNED_CERTIFICATES',)

# base directory for pki storage (should be writable), defaults to PKI_APP_DIR/PKI
PKI_DIR = getattr(settings, 'PKI_DIR', os.path.join(PKI_APP_DIR, 'PKI'))

# path to openssl executable
PKI_OPENSSL_BIN = getattr(settings, 'PKI_OPENSSL_BIN', '/usr/bin/openssl')

# path to generated openssl.conf
PKI_OPENSSL_CONF = getattr(settings, 'PKI_OPENSSL_CONF', os.path.join(PKI_DIR, 'openssl.conf'))

# template name for openssl.conf
PKI_OPENSSL_TEMPLATE = getattr(settings, 'PKI_OPENSSL_TEMPLATE', 'pki/openssl.conf.in')

# jquery url (defaults to pki/jquery-1.3.2.min.js)
JQUERY_URL = getattr(settings, 'JQUERY_URL', 'pki/js/jquery-1.5.min.js')

# logging (TODO: syslog, handlers and formatters)
PKI_LOG = getattr(settings, 'PKI_LOG', os.path.join(PKI_DIR, 'pki.log'))
PKI_LOGLEVEL = getattr(settings, 'PKI_LOGLEVEL', 'debug')

# get other settings directly from settings.py:
ADMIN_MEDIA_PREFIX = getattr(settings, 'ADMIN_MEDIA_PREFIX')

# media url
MEDIA_URL = getattr(settings, 'MEDIA_URL')

# base url: Automatically determined
PKI_BASE_URL = getattr(settings, 'PKI_BASE_URL', get_script_prefix())

# self_signed_serial; The serial a self signed CA starts with. Set to 0 or 0x0 for a random number
PKI_SELF_SIGNED_SERIAL = getattr(settings, 'PKI_SELF_SIGNED_SERIAL', 0x0)

# default key length: The pre-selected key length
PKI_DEFAULT_KEY_LENGTH = getattr(settings, 'PKI_DEFAULT_KEY_LENGTH', 1024)

# default_country: The default country selected (2-letter country code)
PKI_DEFAULT_COUNTRY = getattr(settings, 'PKI_DEFAULT_COUNTRY', 'DE')

# passphrase_min_length: The minimum passphrase length
PKI_PASSPHRASE_MIN_LENGTH = getattr(settings, 'PKI_PASSPHRASE_MIN_LENGTH', 8)

# enable graphviz_support: When True django-pki will render Graphviz PNG's to show relations
PKI_ENABLE_GRAPHVIZ = getattr(settings, 'PKI_ENABLE_GRAPHVIZ', False)

# graphviz direction: From left to right (LR) or top down (TD)
PKI_GRAPHVIZ_DIRECTION = getattr(settings, 'PKI_GRAPHVIZ_DIRECTION', 'LR')

# enable email delivery: Certificates with defined email address can be sent via email
PKI_ENABLE_EMAIL = getattr(settings, 'PKI_ENABLE_EMAIL', False)

